app/src/main/java/com/github/wnder/TakePictureFragment.java

```diff
@@ -267,4 +310,81 @@ private void uploadTakenPicture(String takenPictureId, Uri takenPictureUri) {
             }
         }
     }
+    
+    private void ChooseTourName(){
+        //Hide and show the concerned part of the UI
+        enterText.setVisibility(View.VISIBLE);
+        validateTour.setVisibility(View.VISIBLE);
+        uploadButton.setVisibility(View.INVISIBLE);
+        takePictureButton.setVisibility(View.INVISIBLE);
+        numberOfPictures.setVisibility(View.INVISIBLE);
+        validateTour.setOnClickListener(button -> finalizeTour());
+    }
+
+    private void finalizeTour(){
+        //To avoid empty string as name
+        if(enterText.getText().toString().equals("")){
+            return;
+        }
+        if(user instanceof GuestUser){
+            Snackbar.make(getView(), "Guest user are not allowed to upload tour", Snackbar.LENGTH_SHORT).show();
+        }
+        else{
+            List<String> pictures = new ArrayList<>();
+            for(Pair<String, UploadInfo> pair : tourPictures){
+
+                CompletableFuture<Void> uploadPic = picturesDb.uploadPicture(pair.first, pair.second);
+
+                uploadPic.exceptionally(res -> {
+                    Snackbar.make(getView(), "One or more pictures failed to be uploaded", Snackbar.LENGTH_SHORT).show();
+                    pictures.clear(); //Signify that we lost one picture along the way
+                    return null;
+                }).thenAccept(res -> {
+                    pictures.add(pair.first);
+
+                    if(pictures.size() == tourPictures.size()){ //Which mean that we did not lose any pictures along the way
```

Comment:
else what do we do ? nothing ? we could give the user an alert

Reply:
The goal of this verification is to see when the last picture is uploaded. If it is false, it can mean that the we have not uploaded all the pictures yet (so it's not an error). Anyway, if one picture fail to be upload, the user is informed of that. I will just make the comment clearer.